nupic/research/frameworks/continuous_learning/dendrite_layers.py

```diff
@@ -0,0 +1,153 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from dend_kwinners import DKWinners
+from nupic.torch.modules import SparseWeights  # , KWinnersBase, KWinners
+
+
+class DendriteInput(nn.Module):
+    def __init__(self,
+                 in_dim,
+                 n_dendrites,
+                 threshold=2,
+                 sparse_weights=True,
+                 weight_sparsity=0.2,
+                 #  percent_on=0.1,
+                 #  k_inference_factor=1,
+                 #  boost_strength=2.,
+                 #  boost_strength_factor=0.9,
+                 #  duty_cycle_period=1000,
+                 ):
+        super(DendriteInput, self).__init__()
+        self.threshold = threshold
+        linear = nn.Linear(in_dim, n_dendrites)
+
+        if sparse_weights:
+            self.linear = SparseWeights(linear, weight_sparsity)
+        else:
+            self.linear = linear
+
+        # if 0 < percent_on < 1.0:
+        #     self.act_fun = KWinners(
+        #         n=n_dendrites,
+        #         percent_on=percent_on,
+        #         k_inference_factor=k_inference_factor,
+        #         boost_strength=boost_strength,
+        #         boost_strength_factor=boost_strength_factor,
+        #         duty_cycle_period=duty_cycle_period,
+        #     )
+        # else:
+            # self.act_fun = nn.ReLU()
+
+        # self.act_fun = ADA_fun().cuda()
+
+    def dendrite_activation(self, x):
+        return torch.clamp(x, min=self.threshold)
+
+    def forward(self, x):
+        out = self.linear(x)
+        return out  # self.act_fun(out)
+
+
+class DendriteOutput(nn.Module):
+    def __init__(self, out_dim, dpc):
+        super(DendriteOutput, self).__init__()
+        self.dpc = dpc
+        self.register_buffer("mask", self.dend_mask(out_dim))
+        self.weight = torch.nn.Parameter(torch.Tensor(out_dim, dpc * out_dim))
+        self.bias = torch.nn.Parameter(torch.Tensor(out_dim))
+        nn.init.xavier_normal_(self.weight)
+        self.bias.data.fill_(0.)
+
+    def forward(self, x):
+        w = self.weight * self.mask
+        return F.linear(x, w, self.bias)
+
+    def dend_mask(self, out_dim):
+        mask = torch.zeros(out_dim, out_dim)
+        torch.diag
+        inds = np.diag_indices(out_dim)
+        mask[inds[0], inds[1]] = 1.
+        out_mask = torch.repeat_interleave(mask, self.dpc, dim=0).T
+        return out_mask
+
+
+class DendriteLayer(nn.Module):
```

Comment:
Not sure you need three classes here - this class really doesn't do much. The DendriteOutput class could have a DendriteInput as an instance and do the appropriate forward() logic. You can then get rid of DendriteLayer.

Reply:
I wrote it this way so it would be easier to debug and add features to the different components. Happy to implement your suggestion though if you think that's still preferable

Reply:
OK, leave it as is for now.